File: apps/agent/src/agents/workspaceAgent/tools/addTask.ts
import { FunctionTool } from '@google/adk';
import type { TaskProjection } from '@lifecoach/shared-types';
import { z } from 'zod';
import type { WorkspaceTokensStore } from '../../../storage/workspaceTokens.js';
import type { ExecFileLike, GwsErrorCode } from '../gwsExec.js';
import { type RawTask, projectTask } from '../projections/task.js';
import { type RunGwsLogEvent, runGws } from '../runGws.js';

/**
 * `add_task` — single-step Google Tasks insert.
 */

export const ADD_TASK_TOOL_NAME = 'add_task';

export interface CreateAddTaskToolDeps {
  store: WorkspaceTokensStore;
  uid: string;
  execFile?: ExecFileLike;
  log?: (event: RunGwsLogEvent) => void;
}

export type AddTaskResult =
  | { status: 'ok'; task: TaskProjection }
  | { status: 'error'; code: GwsErrorCode; message: string };

const parameters = z.object({
  title: z.string().min(1).describe('Task title.'),
  due: z
    .string()
    .optional()
    .describe('Optional RFC3339 due date — Google Tasks treats this as a date-only value.'),
  notes: z.string().optional().describe('Optional task notes.'),
  taskListId: z.string().optional().describe('Task list id. Default "@default".'),
});

export function createAddTaskTool(deps: CreateAddTaskToolDeps): FunctionTool {
  const { store, uid, execFile, log } = deps;

  // biome-ignore lint/suspicious/noExplicitAny: zod nominal mismatch with ADK generics
  return new FunctionTool<any>({
    name: ADD_TASK_TOOL_NAME,
    description:
      "Add a task to the user's Google Tasks. Use after surfacing an action from triage_inbox or when the user explicitly asks to add a task. Returns the created task.",
    parameters,
    execute: async (input: unknown): Promise<AddTaskResult> => {
      const args = input as { title: string; due?: string; notes?: string; taskListId?: string };
      const taskListId = args.taskListId ?? '@default';

      const requestBody: Record<string, unknown> = { title: args.title };
      if (args.due) requestBody.due = args.due;
      if (args.notes) requestBody.notes = args.notes;

      const result = await runGws({
        store,
        uid,
        toolName: ADD_TASK_TOOL_NAME,
        service: 'tasks',
        resource: 'tasks',
        method: 'insert',
        params: { tasklist: taskListId },
        body: requestBody,
        execFile,
        log,
      });
      if (result.status === 'error') {
        return { status: 'error', code: result.code, message: result.message };
      }
      const projection = projectTask(
        ((result.body as RawTask | null) ?? {}) as RawTask,
        taskListId,
      );
      return { status: 'ok', task: projection };
    },
  });
}

File: apps/agent_py/src/lifecoach_agent/workspace_agent/tools/add_task.py
"""`add_task` — single-step Google Tasks insert."""

from __future__ import annotations

from typing import Any

from lifecoach_agent.workspace_agent.projections import project_task
from lifecoach_agent.workspace_agent.run_gws import RunGwsOk, run_gws
from lifecoach_agent.workspace_agent.tools._deps import WorkspaceToolDeps

ADD_TASK_TOOL_NAME = "add_task"


def create_add_task_tool(deps: WorkspaceToolDeps) -> Any:
    async def add_task(
        title: str,
        due: str | None = None,
        notes: str | None = None,
        taskListId: str = "@default",  # noqa: N803
    ) -> dict[str, Any]:
        """Add a task to the user's Google Tasks. Use after surfacing an
        action from triage_inbox or when the user explicitly asks to add
        a task. Returns the created task.

        Args:
            title: Task title.
            due: Optional RFC3339 due date — Google Tasks treats this as
                a date-only value.
            notes: Optional task notes.
            taskListId: Task list id. Default "@default".
        """
        request_body: dict[str, Any] = {"title": title}
        if due:
            request_body["due"] = due
        if notes:
            request_body["notes"] = notes

        result = await run_gws(
            store=deps.store,
            uid=deps.uid,
            tool_name=ADD_TASK_TOOL_NAME,
            service="tasks",
            resource="tasks",
            method="insert",
            params={"tasklist": taskListId},
            body=request_body,
            build_client=deps.build_client,
            log=deps.log,
        )
        if not isinstance(result, RunGwsOk):
            return {"status": "error", "code": result.code, "message": result.message}
        raw = result.body if isinstance(result.body, dict) else {}
        projection = project_task(raw, taskListId)
        return {"status": "ok", "task": projection.model_dump(by_alias=True, exclude_none=True)}

    from google.adk.tools import FunctionTool

    add_task.__name__ = ADD_TASK_TOOL_NAME
    return FunctionTool(add_task)
